sql/idempotent/005-vectorizer.sql

```diff
@@ -0,0 +1,1178 @@
+    if grant_to is not null then
+        select pg_catalog.format
+        ( $sql$grant select on %I.%I to %s$sql$
+        , view_schema
+        , view_name
+        , (
+            select pg_catalog.string_agg(pg_catalog.quote_ident(x), ', ')
+            from pg_catalog.unnest(grant_to) x
+          )
+        ) into strict _sql;
+        execute _sql;
+    end if;
+
+    -- schedule the async ext job
+    select ai._vectorizer_schedule_async_ext_job
+    (_vectorizer_id
+    , scheduling
+    ) into _job_id
+    ;
+    if _job_id is not null then
+        scheduling = pg_catalog.jsonb_insert(scheduling, array['job_id'], to_jsonb(_job_id));
+    end if;
+
+    insert into ai.vectorizer
+    ( id
+    , asynchronous
+    , external
+    , source_schema
+    , source_table
+    , source_pk
+    , target_schema
+    , target_table
+    , view_schema
+    , view_name
+    , trigger_name
+    , queue_schema
+    , queue_table
+    , config
+    )
+    values
+    ( _vectorizer_id
+    , asynchronous
+    , external
+    , _source_schema
+    , _source_table
+    , _source_pk
+    , target_schema
+    , target_table
+    , view_schema
+    , view_name
+    , _trigger_name
+    , queue_schema
+    , queue_table
+    , pg_catalog.jsonb_build_object
+      ( 'version', '@extversion@'
+      , 'embedding', embedding
+      , 'chunking', chunking
+      , 'indexing', indexing
+      , 'formatting', formatting
+      , 'scheduling', scheduling
+      )
+    );
+
+    -- grant select on vectorizer table _grant_to roles
+    -- TODO: is there a better way to do this?
+    if grant_to is not null then
+        -- grant select on the users that do NOT already have it
+        for _sql in
+        (
+            select pg_catalog.format($sql$grant select on ai.vectorizer to %I$sql$, x)
+            from pg_catalog.unnest(grant_to) x
+            where not pg_catalog.has_table_privilege
+            ( x
+            , 'ai.vectorizer'
+            , 'select'
+            )
+        )
+        loop
+            execute _sql;
+        end loop;
+    end if;
+
+    -- insert into queue any existing rows from source table
+    select pg_catalog.format
+    ( $sql$
+    insert into %I.%I (%s)
+    select %s
+    from %I.%I x
+    ;
+    $sql$
+    , queue_schema, queue_table
+    , (
+        select pg_catalog.string_agg(pg_catalog.format('%I', x.attname), ', ' order by x.attnum)
+        from pg_catalog.jsonb_to_recordset(_source_pk) x(attnum int, attname name)
+      )
+    , (
+        select pg_catalog.string_agg(pg_catalog.format('x.%I', x.attname), ', ' order by x.attnum)
+        from pg_catalog.jsonb_to_recordset(_source_pk) x(attnum int, attname name)
+      )
+    , _source_schema, _source_table
+    ) into strict _sql
+    ;
+    execute _sql;
+
+    return _vectorizer_id;
+end
+$func$ language plpgsql volatile security invoker
+set search_path to pg_catalog, pg_temp
+;
+
+create or replace function ai.drop_vectorizer(vectorizer_id int) returns void
```

Comment:
can we add a comment about the semantics here? Also, shouldn't drop delete the target table data or did we decide against that?